src/utils.ts

```diff
@@ -1,6 +1,27 @@
 
 import { logger } from './logger';
 
+declare global {
+  interface Array<T> {
+    firstOrDefault(func?: any): T | null;
+  }
+}
+
+Array.prototype.firstOrDefault = function (func?) {
+  if (this.length == 0) {
+    return null;
+  }//from  ww w .  j  av  a2 s .c o  m
+  if (func == undefined || func == null) {
+    return this[0];
+  }
+  this.forEach(element => {
+    if (func.call(this)) {
+      return element;
+    }
```

Comment:
This is wrong: it won't `return` from `firstOrDefault` but from the `forEach` iterator function....

Using a `for(...) {` construct will solve this bug.

However, I see this function is used to access `[0]`... Which is fine to do even on an empty array in JS (only gives `undefined`). So there's no point keeping it here.

Reply:
Also I think `func()` is clearer that `func.call(this)` ;-)